#39382: Fix incorrect image display for directories with identical names but different case

bhushan-cs opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
This pull request addresses an issue where files uploaded to a specific directory in the Media Gallery are also visible in directories with similar names but different case.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#39382

### Manual testing scenarios (*)
1. OS - Ubuntu
2. Navigate to Content > Media Gallery
3. Create a folder under wysiwyg "Testing"
4. Upload an image in the newly created folder.
5. Select wysiwyg folder and try to create another folder named "testing"
6. Image uploaded to "Testing" folder should not show in "testing" folder.


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

bhushan-cs committed:
Media gallery: Fix incorrect image display for directories with identical names but different case

bhushan-cs commented:
@magento run all tests

bhushan-cs committed:
Merge branch 'magento:2.4-develop' into 39382-media-gallery-case-insensitive-bug

bhushan-cs commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into 39382-media-gallery-case-insensitive-bug

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (changes requested):
Hello @bhushan-cs,

Thanks for the contribution!

The code changes seems good to us, but please refer to the below review comments:

1. Add some automated tests for case-sensitive path filtering.
2. Add code comments explaining the BINARY requirement.

bhushan-cs committed:
Merge branch 'magento:2.4-develop' into 39382-media-gallery-case-insensitive-bug

bhushan-cs committed:
Media gallery: Fix incorrect image display for directories with identical names but different case

bhushan-cs commented:
Hello @engcom-Hotel,

Thank you for the feedback!

I have addressed both review comments:
1. Automated Tests Added
I've implemented comprehensive test coverage for case-sensitive path filtering:
Unit Tests: `app/code/Magento/MediaGalleryUi/Test/Unit/Model/SearchCriteria/CollectionProcessor/FilterProcessor/DirectoryTest.php`
Integration Tests: `app/code/Magento/MediaGalleryUi/Test/Integration/Model/SearchCriteria/CollectionProcessor/FilterProcessor/DirectoryIntegrationTest.php`
2. Code Comments Added
Added inline comments which explains why BINARY comparison prevents incorrect file visibility across case-variant directories.

Please let me know if you need any adjustments to the tests or documentation!

engcom-Hotel commented on app/code/Magento/MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/Directory.php:
```diff
@@ -19,7 +19,16 @@ class Directory implements CustomFilterInterface
+         * The regex '^{path}/[^\/]*$' ensures we only match files directly in the
+         * specified directory, not in subdirectories.
+         */
+        $collection->getSelect()->where('BINARY path REGEXP ? ', '^' . $value . '/[^\/]*$');
```
Suggestion: Consider to add fallback behaviour with error handling:
```suggestion
       try {
    $collection->getSelect()->where('BINARY path REGEXP ? ', '^' . $value . '/[^\/]*$');
} catch (\Exception $e) {
    // Log error and fallback to non-BINARY comparison
    $this->logger->error('BINARY regex failed, falling back: ' . $e->getMessage());
    $collection->getSelect()->where('path REGEXP ? ', '^' . $value . '/[^\/]*$');
}
```

bhushan-cs replied:
Hello @engcom-Hotel,

The static test is giving warning on Magento copywriter comment. Do we have to change it to B2B copywriter comment?
```php
* Copyright 2019 Adobe
* All Rights Reserved.
 ```
 

> Suggestion: Consider to add fallback behaviour with error handling:

Great suggestion! I agree that adding error handling makes the solution more robust. I've added a few small modifications:
```php
try {
    $collection->getSelect()->where('BINARY path REGEXP ? ', '^' . $value . '/[^\/]*$');
} catch (\Exception $e) {
    // Log the error for debugging but continue with case-insensitive fallback
    // Note: This fallback means directory filtering will not be case-sensitive
    $this->logger->error('MediaGallery Directory Filter: BINARY REGEXP not supported, using case-insensitive fallback: ' . $e->getMessage());
    $collection->getSelect()->where('path REGEXP ? ', '^' . $value . '/[^\/]*$');
}
```
I was not sure how to edit the code suggestion to add my changes.

engcom-Hotel replied:
> The static test is giving warning on Magento copywriter comment. Do we have to change it to B2B copywriter comment?

You need to refer the files which is mentioned in the failed test, for example, consider the following screenshot:

<img width="2014" height="516" alt="image" src="https://github.com/user-attachments/assets/41a98b29-d664-48df-a20a-acf79aba1018" />

The above screenshot is from the below static test failure:

https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39502/fa70983fcf29df8efaba5329f6467951/Statics/report-sanity-ce.html

engcom-Hotel replied:
> I was not sure how to edit the code suggestion to add my changes.

You can directly edit this file from your local in the same branch, and commit as usual.

bhushan-cs replied:
Thanks for the guidance. Let me commit the changes.

engcom-Hotel reviewed (changes requested):
Hello @bhushan-cs,

Please fix the failed static test, other tests seems flaky to me.

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

engcom-Dash commented:
Hi @bhushan-cs 

Just checking in to see if you're currently working on the changes. If not, I’m happy to take care of them to help move things forward. Let me know what works best for you.

Thanks!

bhushan-cs committed:
Merge branch 'magento:2.4-develop' into 39382-media-gallery-case-insensitive-bug

bhushan-cs committed:
Added fallback in case BINARY is not supported and fixed copyright message

bhushan-cs commented:
@magento run all tests

bhushan-cs committed:
Fix static test warnings

engcom-Hotel commented:
Hello @bhushan-cs,

It seems the PR has been ready for review, hence picking it for the same.

Thanks

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

engcom-Hotel commented on app/code/Magento/MediaGalleryUi/Test/Unit/Model/SearchCriteria/CollectionProcessor/FilterProcessor/DirectoryTest.php:
```diff
@@ -0,0 +1,175 @@
+<?php
+/**
+ * Copyright 2019 Adobe
```
```suggestion
 * Copyright 2025 Adobe
```

engcom-Hotel commented on app/code/Magento/MediaGalleryUi/Test/Integration/Model/SearchCriteria/CollectionProcessor/FilterProcessor/DirectoryIntegrationTest.php:
```diff
@@ -0,0 +1,212 @@
+<?php
+/**
+ * Copyright 2019 Adobe
```
```suggestion
 * Copyright 2025 Adobe
```

engcom-Hotel commented on app/code/Magento/MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/Directory.php:
```diff
@@ -10,16 +10,49 @@
+    /**
+     * @param LoggerInterface $logger
+     */
+    public function __construct(LoggerInterface $logger)
```
Make logger parameter optional with ObjectManager fallback for backward compatibility

bhushan-cs replied:
I've made the suggested change

engcom-Hotel reviewed (changes requested):
Hello @bhushan-cs,

Thanks for working on the changes!

The failed tests seems flaky to me, hence re-running the tests. Also please refer to the below review comments.

Thanks

bhushan-cs committed:
Update app/code/Magento/MediaGalleryUi/Test/Unit/Model/SearchCriteria/CollectionProcessor/FilterProcessor/DirectoryTest.php

Co-authored-by: Abhinav Pathak <51681618+engcom-Hotel@users.noreply.github.com>

bhushan-cs committed:
Update app/code/Magento/MediaGalleryUi/Test/Integration/Model/SearchCriteria/CollectionProcessor/FilterProcessor/DirectoryIntegrationTest.php

Co-authored-by: Abhinav Pathak <51681618+engcom-Hotel@users.noreply.github.com>

bhushan-cs committed:
Made logger parameter optional with ObjectManager fallback for backward compatibility

bhushan-cs commented:
@magento run all tests

bhushan-cs committed:
Explicitly mark logger parameter as nullable

engcom-Hotel committed:
Merge branch '2.4-develop' into 39382-media-gallery-case-insensitive-bug

engcom-Hotel committed:
Merge branch '2.4-develop' into 39382-media-gallery-case-insensitive-bug

engcom-Hotel commented:
@magento run all tests

bhushan-cs commented:
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, WebAPI Tests

engcom-Hotel reviewed (approved):
Failed tests seems flaky to me, hence approving the PR.

engcom-Hotel committed:
Merge branch '2.4-develop' into 39382-media-gallery-case-insensitive-bug

engcom-Bravo commented:
@magento run all tests

engcom-Dash committed:
Merge branch '2.4-develop' into 39382-media-gallery-case-insensitive-bug

engcom-Dash commented:
@magento run all tests
